### ISSUE TITLE
UI Automation can cause unbounded memory consumption

### ISSUE THREAD

**Issue by filipnavara** [dotnet/winforms#3182]
* .NET Core Version:  3.1.200

* Have you experienced this same bug with .NET Framework?: Yes

**Problem description:**

The WinForms control expose UI Automation interfaces through WM_GETOBJECT. They, however, don't call the `UiaDisconnectProvider` API when the controls are destroyed ([SDK sample](https://github.com/microsoft/Windows-classic-samples/tree/master/Samples/UIAutomationCleanShutdown)). It [does call `UiaReturnRawElementProvider`](https://docs.microsoft.com/en-us/windows/win32/api/uiautomationcoreapi/nf-uiautomationcoreapi-uiareturnrawelementprovider) as specified in the documentation but that does not release any providers that are still held by the UI Automation clients. The clients can still hold on to the COM objects and thus keep the lifetime much longer than the actual lifetime of the controls and forms. This is problematic because it can cause massive memory consumption because the controls and forms cannot get garbage collected. 

The old way of handling this was breaking up the references between the automation proxies and the actual controls and returning UIA_E_ELEMENTNOTAVAILABLE error from the COM entry points. This could possibly be implemented in quite transparent way in the [`InternalAccessibleObject` class](https://github.com/dotnet/winforms/blob/a461d30169401558eef89aba8f3fa04501fd305d/src/System.Windows.Forms/src/System/Windows/Forms/AccessibleObject.cs#L2520).

The newer way of handling this is asking UIA to release any references to the interface when the form/control is destroyed using the `UiaDisconnectProvider` API.

The behavior can be checked by the VisualUIAVerifyNative.exe tool that is part of the Windows SDK.

**Examples:**

Label control on a live form in UIA Verify:

![image](https://user-images.githubusercontent.com/1764393/80861479-1d5b4280-8c6f-11ea-8f91-5e9a284425bf.png)

Label control on a closed form in UIA Verify:

![image](https://user-images.githubusercontent.com/1764393/80861486-2cda8b80-8c6f-11ea-8c7e-74c37b6e10c0.png)

Different control in UIA Verify after the form was closed and `UiaDisconnectProvider` was called:

![image](https://user-images.githubusercontent.com/1764393/80861496-424fb580-8c6f-11ea-9194-ddfaac8f6f30.png)

**Expected behavior:**

UI Automation clients should not leak memory through the server-side UIA providers on disposed controls.

**Comment by RussKie** [comment]
Thank you for the report. 
You seemed to know quite a bit about this area. Do you think you could send a PR with a fix?

/cc: @M-Lipin

**Comment by filipnavara** [comment]
I am looking into it and maybe I will come up with a PR. The documentation on the topic is quite scarce unfortunately.

**Comment by filipnavara** [comment]
Unfortunately the easy fix solves only half of the problem. Calling `UiaDisconnectProvider` helps with all the controls that return `true` from `SupportsUiaProviders`. However, that excludes `UserControl` and anything delivered from it. Since the property is non-public there's no easy way to override it from user code (aside from overriding the whole `WndProc` logic for `WM_GETOBJECT`). The issue basically boils down to the fact that UiaCore has backward compatibility with MSAA's `IAccessibleEx`. If the particular control doesn't return UIA object in WM_GETOBJECT then it falls back to wrapping the MSAA's object for the UIA clients. WinForms then never releases the UIA objects and we are back with the original problem.

I assume the intended way to address this is to use UIA exclusively. It provides a backward compatibility layer for MSAA. If `WM_GETOBJECT` implementation doesn't check `LParam` and goes straight to `UiaReturnRawElementProvider` call then UIA will handle requests for MSAA objects and use its own compatibility layer. Unfortunately that may be a change that would break user code and the lack of documentation in this area does not help.

Here are links to the branches with the `UiaDisconnectProvider` fix:
https://github.com/filipnavara/winforms/tree/uiadisconnect (master)
https://github.com/filipnavara/winforms/tree/uiadisconnect_3.1 (release/3.1)

I'm not going to submit it as PR until there is a consensus on how to solve the issue for MSAA accessibility objects.

**Comment by merriemcgaw** [comment]
Out of curiosity, how is this represented in the [aka.ms/accessibilityinsights](https://aka.ms/accessibilityinsights) tool? That is fast becoming our standard tool for evaluation of accessibility.

**Comment by filipnavara** [comment]
I gave Accessibility Insights a try. Unlike other tools it actually does some proper tracking and live refreshes until you pause it. To reproduce the problematic scenario I opened the form, selected element on the form in the Accessibility Insights and paused it. Then I closed the form.

For things like `Label` you can see that it still allows to get an information of a control that is disposed:

![ai_label](https://user-images.githubusercontent.com/1764393/81441003-40a35780-9171-11ea-9e82-64bbd8c22d6b.png)

For controls where `UiaDisconnectProvider` was called it is empty:

![image](https://user-images.githubusercontent.com/1764393/81441049-59137200-9171-11ea-8267-f07281899e32.png)

The original issue that we had was triggered by the Ranorex UI testing tool. It's notorious in not releasing the UI Automation objects (technically a bug on their end) but the end result is a huge leak in our application that eventually gets so bad that it results in out-of-memory situation. We'd expect some level of resiliency for this so 3rd-party applications would not be able to crash ours through the UI Automation API.

Unfortunately, Accessibility Insights also managed to trigger some scenario where it leaks even in cases which we thought were solved. I'll have to look into it but the memory profiler shows way too many things holding the closed form:

![image](https://user-images.githubusercontent.com/1764393/81441448-1736fb80-9172-11ea-88cd-81f61e32cdf2.png)

**Comment by RussKie** [comment]
Thank you for the investigations.

> I'm not going to submit it as PR until there is a consensus on how to solve the issue for MSAA accessibility objects.

It is easier to review changes, discuss and bounce idea in a PR. First we need to fix it in the master branch, and then put out a case for patching 3.1 branch.
Something else perhaps to consider is calling [`UiaDisconnectAllProviders`](https://docs.microsoft.com/windows/win32/api/uiautomationcoreapi/nf-uiautomationcoreapi-uiadisconnectallproviders) when shutting the app.

**Comment by filipnavara** [comment]
> It is easier to review changes, discuss and bounce idea in a PR. First we need to fix it in the master branch, and then put out a case for patching 3.1 branch.

#3308. I shared the backport to release/3.1 only because I already had it for testing on our own app. It's non-trivial for us to retarget the app to 5.0 so any large scale testing with memory profiler I do with the 3.1 backports on actual production version of the app.

> Something else perhaps to consider is calling UiaDisconnectAllProviders when shutting the app.

That's pretty much useless when the app is shutting down because the COM objects will die anyway. `UiaDisconnectAllProviders` is meant mostly for the UIA clients, not the providers, as I understand it.

For UIA providers it's recommended to use `UiaDisconnectProvider` to support proper unloading. The older MSAA API seems to have no direct alternative unfortunately. One can call `CoDisconnectObject` on the object that implements `IAccessible` (ie. `InternalAccessibleObject`) presumably but I have no idea what kind of compatibility problems that could introduce. Presumably `UiaDisconnectProvider` may call that itself under the hood. If that's the case I would expect it to do it also for the bridged MSAA objects (currently not used by WinForms) and hence it may not actually be a compatibility concern.

**Comment by RussKie** [comment]
@filipnavara how much of an issue is this for you if the fix isn't ported to 3.1?

**Comment by filipnavara** [comment]
> @filipnavara how much of an issue is this for you if the fix isn't ported to 3.1?

We are evaluating the impact at the moment with a group of our beta testers.

It is certainly a blocker for us for using our automated testing (using the Ranorex tool). That's something we can live with by splitting the tests into small chunks and terminating the test process between them.

The bigger problem would be if it caused unintentional out-of-memory errors for customers. Unfortunately we have really poor diagnostics story to analyze this kind of issues and to determine the root cause. So far it looks that the common screen reader software didn't trigger this issue yet. However, we were surprised that on a small sample of our beta testers (few thousands) we see about 4% of them having some software using UI Automation on their machines. In some cases we saw UIA calls even if the customer claimed not to use any screen reader software. That's some data that makes me a bit uneasy about shipping our product with the issue.

**Comment by RussKie** [comment]
@merriemcgaw

**Comment by weltkante** [comment]
>  In some cases we saw UIA calls even if the customer claimed not to use any screen reader software.

When I've been debugging WinForms going through automation API was the common case not the exception. Windows itself is very happy to use UI automation for various purposes, for example the onscreen keyboard will go through the automation API, there are probably more common tools that use it but I'm not actively aware of them.
